packages/algolia-experiences/src/render.tsx

```diff
@@ -0,0 +1,252 @@
+  if (text.type === 'string') {
+    return text.value;
+  }
+
+  if (text.type === 'attribute') {
+    return getPropertyByPath(hit, text.path);
+  }
+
+  return null;
+}
+
+function childToWidget(child: Child, container: HTMLElement): Widget[] {
+  const widgetContainer = container.appendChild(document.createElement('div'));
+
+  if (child.type === 'columns') {
+    widgetContainer.classList.add('ais-Columns');
+
+    return child.children
+      .map((column) => {
+        const columnContainer = widgetContainer.appendChild(
+          Object.assign(document.createElement('div'), {
+            className: 'ais-Column',
+          })
+        );
+        return column.map((ch) => childToWidget(ch, columnContainer));
+      })
+      .flat(2);
+  }
+
+  if (child.type === 'ais.configure') {
+    return [widgets[child.type]({ ...child.parameters })];
+  }
+
+  if (isTemplateWidget(child)) {
+    // type cast is needed here because the spread adding `container` and `templates` loses the type discriminant
+    const parameters = child.parameters as Parameters<
+      typeof widgets['ais.hits']
+    >[0];
+    const widget = widgets[child.type] as typeof widgets['ais.hits'];
+
+    return [
+      widget({
+        ...parameters,
+        container: widgetContainer,
+        templates: {
+          item: (hit: any, { components }) => {
+            if (!child.children.length) {
+              return <code> no item template given</code>;
+            }
+
+            function renderChild(ch: TemplateChild) {
+              const Tag = tagNames.get(ch.type) as keyof JSX.IntrinsicElements;
+              if (!Tag) {
+                return <Fragment></Fragment>;
+              }
+
+              let children: ComponentChildren = null;
+              if ('text' in ch.parameters) {
+                children = ch.parameters.text.map((text) =>
+                  renderText(text, hit, components)
+                );
+              } else if ('children' in ch) {
+                children = ch.children.map(renderChild);
+              }
+
+              const attributes = Object.fromEntries(
+                Object.entries(ch.parameters)
+                  .filter(
+                    (tuple): tuple is [string, TemplateAttribute] =>
+                      tuple[0] !== 'text'
+                  )
+                  .map(([key, value]) => [
+                    key,
+                    value.map((item) => renderAttribute(item, hit)).join(''),
+                  ])
+              );
+
+              return <Tag {...attributes}>{children}</Tag>;
+            }
+
+            return child.children.map(renderChild);
+          },
+        },
+      }),
+    ];
+  }
+
+  if (isPanelWidget(child)) {
+    // type cast is needed here because the spread adding `container` loses the type discriminant
+    const {
+      header,
+      collapsed: defaultCollapsed,
+      ...parameters
+    } = child.parameters as Parameters<
+      typeof widgets['ais.refinementList']
```

Comment:
Is it casted specifically as Refinement List because we're focusing only on this widget for the pilot?

Reply:
the type parameters are lost (because of the spread including container I believe), so they need to be casted as one specific widget that has those parameters